packages/zoe/src/contractSupport/recorder.js

```diff
@@ -0,0 +1,264 @@
+import { Fail } from '@agoric/assert';
+import { makeTypeGuards } from '@agoric/internal';
+import { prepareDurablePublishKit } from '@agoric/notifier';
+import { mustMatch } from '@agoric/store';
+import { M, prepareExoClass } from '@agoric/vat-data';
+import { E } from '@endo/eventual-send';
+
+// XXX UNTIL https://github.com/Agoric/agoric-sdk/issues/7090
+const { StorageNodeShape } = makeTypeGuards(M);
+
+/**
+ * Recorders support publishing data to vstorage.
+ *
+ * `Recorder` is similar to `Publisher` (in that they send out data) but has different signatures:
+ * - methods are async because they await remote calls to Marshaller and StorageNode
+ * - method names convey the durability
+ * - omits fail()
+ * - adds getStorageNode() from its durable state
+ *
+ * Other names such as StoredPublisher or ChainStoragePublisher were considered, but found to be sometimes confused with *durability*, another trait of this class.
+ */
+
+/**
+ * @template T
+ * @typedef {{ getStorageNode(): StorageNode, getStoragePath(): Promise<string>, write(value: T): Promise<void>, writeFinal(value: T): Promise<void> }} Recorder
+ */
+
+/**
+ * @template T
+ * @typedef {Pick<PublishKit<T>, 'subscriber'> & { recorder: Recorder<T> }} RecorderKit
+ */
+
+/**
+ * @template T
+ * @typedef {Pick<PublishKit<T>, 'subscriber'> & { recorderP: ERef<Recorder<T>> }} EventualRecorderKit
+ */
+
+/**
+ * Wrap a Publisher to record all the values to chain storage.
+ *
+ * @param {import('@agoric/zoe').Baggage} baggage
+ * @param {ERef<Marshaller>} marshaller
+ */
+export const prepareRecorder = (baggage, marshaller) => {
+  /**
+   */
+  const makeRecorder = prepareExoClass(
+    baggage,
+    'Recorder',
+    M.interface('Recorder', {
+      getStorageNode: M.call().returns(StorageNodeShape),
+      getStoragePath: M.call().returns(M.promise(/* string */)),
+      write: M.call(M.any()).returns(M.promise()),
+      writeFinal: M.call(M.any()).returns(M.promise()),
+    }),
+    /**
+     * @template T
+     * @param {PublishKit<T>['publisher']} publisher
+     * @param {StorageNode} storageNode
+     * @param {TypedMatcher<T>} [valueGuard]
+     */
+    (publisher, storageNode, valueGuard = M.any()) => {
+      return {
+        closed: false,
+        publisher,
+        storageNode,
+        storagePath: /** @type {string | undefined} */ (undefined),
+        valueGuard,
```

Comment:
```suggestion
     * @param {TypedMatcher<T>} [valueShape]
     */
    (publisher, storageNode, valueShape = M.any()) => {
      return {
        closed: false,
        publisher,
        storageNode,
        storagePath: /** @type {string | undefined} */ (undefined),
        valueShape,
```
(per https://github.com/Agoric/agoric-sdk/blob/f4a34abadce17db3f68d611419cb41bdc3f4f4d9/packages/store/src/types.js#L54-L60 and https://github.com/endojs/endo/blob/master/packages/patterns/README.md )